### PR TITLE
fix: Policies should apply only to gateways they were attached to when mergeGateways is true

### LINKED ISSUE #2666: When there are multiple Gateways and `mergeGateway `is enabled , using CORS on one Gateway will affect the others
*Description*:
When there are multiple Gateways and `mergeGateway` is enabled , using CORS on one Gateway will affect the others 

*Repro steps*:
Install the Gateway API CRDs and Envoy Gateway 
```shell
helm install eg oci://docker.io/envoyproxy/gateway-helm --version v0.0.0-latest -n envoy-gateway-system --create-namespace
```

Use aill-in-one yaml described below:

```shell
apiVersion: gateway.networking.k8s.io/v1
kind: GatewayClass
metadata:
  name: eg
spec:
  controllerName: gateway.envoyproxy.io/gatewayclass-controller
  parametersRef:
    group: gateway.envoyproxy.io
    kind: EnvoyProxy
    name: custom-proxy-config
    namespace: envoy-gateway-system

---
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  name: custom-proxy-config
  namespace: envoy-gateway-system
spec:
  mergeGateways: true 
---
apiVersion: v1
kind: Namespace
metadata:
  name: sealos-system
---
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: eg
  namespace: sealos-system
spec:
  gatewayClassName: eg
  listeners:
    - hostname: "*.192.168.0.15.nip.io"
      name: http
      protocol: HTTP
      port: 80
      allowedRoutes:
        namespaces:
          from: All
---
apiVersion: v1
kind: Namespace
metadata:
  name: ns-oehe
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: bdkzlmibsivuiqav
  namespace: ns-oehe
spec:
  replicas: 1
  selector:
    matchLabels:
      app: bdkzlmibsivuiqav
  template:
    metadata:
      labels:
        app: bdkzlmibsivuiqav
    spec:
      containers:
      - image: nginx:1.25
        name: nginx
        ports:
        - containerPort: 80
          name: http
          protocol: TCP
        resources:
          limits:
            cpu: 100m
            memory: 64Mi
          requests:
            cpu: 10m
            memory: 16Mi
---
apiVersion: v1
kind: Service
metadata:
  name: bdkzlmibsivuiqav
  namespace: ns-oehe
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: bdkzlmibsivuiqav
  type: ClusterIP
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: bdkzlmibsivuiqav
  namespace: ns-oehe
spec:
  hostnames:
  - ntjxuedx.192.168.0.15.nip.io
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
    namespace: sealos-system
    sectionName: http
  rules:
  - backendRefs:
    - group: ""
      kind: Service
      name: bdkzlmibsivuiqav
      port: 80
      weight: 1
    matches:
    - path:
        type: PathPrefix
        value: /
---
apiVersion: v1
kind: Namespace
metadata:
  name: ns-vrpg
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mfqjpuycbgjrtdww
  namespace: ns-vrpg
spec:
  replicas: 1
  selector:
    matchLabels:
      app: mfqjpuycbgjrtdww
  template:
    metadata:
      labels:
        app: mfqjpuycbgjrtdww
    spec:
      containers:
      - image: nginx:1.25
        name: nginx
        ports:
        - containerPort: 80
          name: http
          protocol: TCP
        resources:
          limits:
            cpu: 100m
            memory: 64Mi
          requests:
            cpu: 10m
            memory: 16Mi
---
apiVersion: v1
kind: Service
metadata:
  name: mfqjpuycbgjrtdww
  namespace: ns-vrpg
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: mfqjpuycbgjrtdww
  type: ClusterIP
---
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: mfqjpuycbgjrtdww
  namespace: ns-vrpg
spec:
  gatewayClassName: eg
  listeners:
  - hostname: qccbahgo.qccbahgo
    name: http
    port: 80
    protocol: HTTP
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: mfqjpuycbgjrtdww
  namespace: ns-vrpg
spec:
  hostnames:
  - qccbahgo.qccbahgo
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: mfqjpuycbgjrtdww
    namespace: ns-vrpg
    sectionName: http
  rules:
  - backendRefs:
    - group: ""
      kind: Service
      name: mfqjpuycbgjrtdww
      port: 80
      weight: 1
    matches:
    - path:
        type: PathPrefix
        value: /
---
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: cors-example
  namespace: sealos-system
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  cors:
    allowOrigins:
    - "http://*.foo.com"
    allowMethods:
    - PUT
    - GET
    - POST
    - DELETE
    - PATCH
    - OPTIONS
    maxAge: 600s
    allowCredentials: true
```

Get the name of the Envoy service created the by the  Gateway, because mergeGateway is enabled, the way to obtain the Envoy service differs from the quickstart.

```shell
export ENVOY_SERVICE=$(kubectl get svc -n envoy-gateway-system --selector=gateway.envoyproxy.io/owning-gatewayclass=eg -o jsonpath='{.items[0].metadata.name}')
```
Port forward to the Envoy service
```shell
kubectl -n envoy-gateway-system port-forward service/${ENVOY_SERVICE} 8888:80 &
```
Verify that the CORS headers are present in the response of the OPTIONS request from `http://www.foo.com`, and the host is `ntjxuedx.192.168.0.15.nip.io`

```shell
curl -H "Origin: http://www.foo.com" \
  -H "Host: ntjxuedx.192.168.0.15.nip.io" \
  -H "Access-Control-Request-Method: GET" \
  -X OPTIONS -v -s \
  http://localhost:8888/
```
The following is the response, which is reasonable because this host is bound to the Gateway in the targetRef.

```shell
< HTTP/1.1 200 OK
< access-control-allow-origin: http://www.foo.com
< access-control-allow-credentials: true
< access-control-allow-methods: PUT, GET, POST, DELETE, PATCH, OPTIONS
< access-control-max-age: 600
```

Verify that the CORS headers are present in the response of the OPTIONS request from `http://www.foo.com`, and the host is `qccbahgo.qccbahgo`

```shell
curl -H "Origin: http://www.foo.com" \
  -H "Host: qccbahgo.qccbahgo" \
  -H "Access-Control-Request-Method: GET" \
  -X OPTIONS -v -s \
  http://localhost:8888/
```
The following is the response, the same as above which is unreasonable because this host is unrelated to the Gateway in the targetRef.

```shell
< HTTP/1.1 200 OK
< access-control-allow-origin: http://www.foo.com
< access-control-allow-credentials: true
< access-control-allow-methods: PUT, GET, POST, DELETE, PATCH, OPTIONS
< access-control-max-age: 600
```

I think it is unreasonable. Our use case is similar to the following structure: users can only operate on CRs in their own namespaces. However, when using CORS, it affects other Gateways in the cluster, which poses significant risks.

img_v3_0275_843b3e72-c935-451c-8576-b575e7c8069g

### PR DESCRIPTION
**Pull request by liorokman**
**Which issue(s) this PR fixes**:
Fixes #2666

### CODE REVIEW

**Review by cnvergence — APPROVED** [review]
LGTM

File: internal/gatewayapi/backendtrafficpolicy.go [hunk 1]
```diff
@@ -345,7 +345,14 @@ func (t *Translator) translateBackendTrafficPolicyForGateway(policy *egv1a1.Back
 	// Should exist since we've validated this
 	ir := xdsIR[irKey]
 
+	policyTarget := irStringKey(
```
Comment (arkodg) [inline internal/gatewayapi/backendtrafficpolicy.go:348]: can we unify the logic and style used for route matching as well ?
https://github.com/envoyproxy/gateway/blob/c30d037a0018ddb131cacb2f041ae5bc81dcae10/internal/gatewayapi/backendtrafficpolicy.go#L284

with comments on - how prefix is generated and a helper match to compute a match

File: internal/gatewayapi/listener.go [hunk 2]
```diff
@@ -100,10 +100,11 @@ func (t *Translator) ProcessListeners(gateways []*GatewayContext, xdsIR XdsIRMap
 			switch listener.Protocol {
 			case gwapiv1.HTTPProtocolType, gwapiv1.HTTPSProtocolType:
 				irListener := &ir.HTTPListener{
-					Name:    irHTTPListenerName(listener),
-					Address: "0.0.0.0",
-					Port:    uint32(containerPort),
-					TLS:     irTLSConfigs(listener.tlsSecrets),
+					GatewayName: irStringKey(gateway.Namespace, gateway.Name),
```
Comment (arkodg) [inline internal/gatewayapi/listener.go:103]: if we use the `prefix` method approach we can do w/o the extra field and same on memory

File: internal/gatewayapi/backendtrafficpolicy.go (hunk 1 above)
Comment (arkodg) [reply internal/gatewayapi/backendtrafficpolicy.go:348]: https://github.com/envoyproxy/gateway/blob/c30d037a0018ddb131cacb2f041ae5bc81dcae10/internal/gatewayapi/helpers.go#L361

**Comment by liorokman** [comment]
> thanks for fixing this one @liorokman, isn't a similar fix needed in ClientTrafficPolicy as well ?

The way that `ClientTrafficPolicy` works, it filters the relevant gateways according to the `Gateway` resources correctly. It already isolated policies to the correct listeners.

I've updated the unit-test to include all the different types of policies to make sure this behaviour isn't lost.

File: internal/gatewayapi/backendtrafficpolicy.go (hunk 1 above)
Comment (liorokman) [reply internal/gatewayapi/backendtrafficpolicy.go:355]: Sorry, I don't follow your meaning.

File: internal/gatewayapi/listener.go (hunk 2 above)
Comment (liorokman) [reply internal/gatewayapi/listener.go]: Done.

**Review by guydc — APPROVED** [review]
LGTM

**Review by arkodg — APPROVED** [review]
LGTM

**Pull request merged**